src/Env/EnvFactory.php

```diff
@@ -24,13 +25,36 @@
  */
 class EnvFactory
 {
-    public static function createEnv(EnvEnum $type, AppName $appName, \Traversable $dependencies): Env
+    public static function createEnv(EnvEnum $name, AppName $appName, \Traversable $dependencies): Env
     {
         return new Env(
-            (string) $type,
-            new Vagrant($type, $appName, VagrantBoxVersionResolver::resolve($dependencies)),
-            new Ansible($type, ...$dependencies),
-            new Make($type)
+            (string) $name,
+            new Vagrant($name, $appName, VagrantBoxVersionResolver::resolve($dependencies)),
+            new Ansible($name, ...$dependencies),
+            new Make($name)
         );
     }
+
+    public static function createEnvFromMetadata(string $rawName, array $metadata): Env
+    {
+        $configs = [];
+        $name = EnvEnum::create($rawName);
+        $hydrator = new VariableHydrator();
+
+        foreach ($metadata as $configClass => $perClassVars) {
+            $hydratedVars = [];
+
+            foreach ($perClassVars as $class => $vars) {
+                foreach ($vars as $data) {
+                    $var = (new \ReflectionClass($class))->newInstanceWithoutConstructor();
+                    $hydrator->hydrate($var, $data);
+                    $hydratedVars[] = $var;
+                }
+            }
+
+            $configs[] = new $configClass($name, ...$hydratedVars);
```

Comment:
> make Config::__construct() final or define a named constructor which would be final too, right?

Making the `__construct` final could be an option. But I would avoid it, as it'll prevent doing anything in the constructor in a child class if we need it in the future.

The named constructor doesn't need to be final, as it has to respect exactly the signature (that's the only think we need) and still allow to override it. It can still define extra arguments with defaults, but I don't think we should care.